django/db/models/expressions.py

```diff
@@ -902,7 +902,21 @@ def resolve_expression(
         return query.resolve_ref(self.name, allow_joins, reuse, summarize)
 
     def replace_expressions(self, replacements):
-        return replacements.get(self, self)
+        if (replacement := replacements.get(self)) is not None:
+            return replacement
+        field_name, *transforms = self.name.split(LOOKUP_SEP)
+        if not transforms:
+            return self
+        if (
+            replacement := replacements.get(F(field_name))
+        ) is None or replacement._output_field_or_none is None:
```

Comment:
The `_output_field_or_none` check is important as `BaseExpression.get_transform` delegates to `self.output_field`. The alternative would be to catch `OutputFieldIsNoneError` when calling `get_transform` below but that's wasteful when dealing with nested transforms.